reactioncommerce/reaction
# feat: Create product hash of published product properties

Resolves #4264 
Impact: **minor**  
Type: **feature**

## Issue
After the removal of revisions, we need a way to allow an admin user to see if a product has been updated. Saving a hash of the data in it's current state, to allow for comparison, will allow the admin user to see if changes have been made.

## Solution
Use the `object-hash` npm library to create a hash based on specific fields of a product that we want to track. Save this hash to the product itself, to check against to see if fields have been updated (ticket #4281)

Most fields are included in this hash. The primary excluded fields are system generated auto-fields such as `createdAt` and `updatedAt`, and the `hash` itself.

Additional notes:
- Moved the function that created the product data to be saved to the Catalog into it's own file, `createCatalogProduct`, so we can keep it separate incase we ever want to create / see the catalog data without publishing it at the same time. 
- Added a migration to add the hash to products which have already been published. Although migration 24 already will add the hash, it's already been released to master so this is needed for anyone past 24.

## Breaking changes
The function `createCatalogProduct` has been moved into it's own file. This _should not_ create any issues, as it was not an exported function, but just be aware.

## Testing
1. Open Robo3T to see the Products and Catalog collections
2. Create / edit a product
3. See the hash update inside both collections

OR

Start the app and see the hash on a product that's already created, via migration.
